File: frontend/lib/types/smart-review.ts
/**
 * TypeScript types for Smart Adaptive Review Editor
 *
 * Matches backend Pydantic schemas in app/schemas/review_slot.py
 */

// ===== Phase 1: Quick Assessment =====

export interface Phase1QuickAssessment {
  overall_rating: number; // 1-5
  primary_focus_areas: string[]; // e.g., ['functionality', 'security']
  quick_summary: string; // 50-300 chars
}

// ===== Phase 2: Rubric Ratings =====

// Rating justification - explains WHY a rating was given
export interface RatingRationale {
  strengths: string; // What earned this score
  gaps: string; // What's holding it back (why not 5?)
}

export interface Phase2RubricRatings {
  content_type: string; // 'code' | 'design' | 'writing'
  ratings: Record<string, number>; // e.g., { functionality: 5, code_quality: 4 }
  rationales?: Record<string, RatingRationale>; // Justification for each rating
}

// ===== Phase 3: Detailed Feedback =====

export interface VisualAnnotation {
  id: string; // UUID generated on client
  x: number; // X coordinate as percentage (0-100)
  y: number; // Y coordinate as percentage (0-100)
  comment: string; // Annotation comment (1-500 chars)
}

export interface VoiceMemo {
  id: string; // UUID generated on client
  duration: number; // Duration in seconds
  url?: string; // Blob URL for playback (client-side only)
  file_url?: string; // Uploaded file URL (for persistence)
}

// Structured feedback for actionable improvements
export type FeedbackPriority = 'critical' | 'important' | 'nice-to-have';

// Premium fields for expert reviews
export type EffortEstimate = 'quick-fix' | 'moderate' | 'major-refactor';
export type ConfidenceLevel = 'certain' | 'likely' | 'suggestion';
export type ImprovementCategory = 'performance' | 'ux' | 'security' | 'accessibility' | 'maintainability' | 'design' | 'content' | 'other';

export interface ResourceLink {
  url: string;
  title?: string;
}

// Common design/development principles that can be referenced
export type PrincipleCategory =
  | 'ux-heuristic' // Nielsen's heuristics, etc.
  | 'design-principle' // Gestalt, color theory, typography
  | 'coding-standard' // SOLID, DRY, clean code
  | 'accessibility' // WCAG, a11y guidelines
  | 'performance' // Core Web Vitals, optimization
  | 'security' // OWASP, security best practices
  | 'seo' // Search engine guidelines
  | 'content' // Writing frameworks, clarity
  | 'other';

// Impact types for explaining consequences
export type ImpactType =
  | 'conversion' // Affects sales/signups
  | 'usability' // Makes harder to use
  | 'trust' // Reduces credibility
  | 'performance' // Slows down experience
  | 'maintainability' // Creates tech debt
  | 'accessibility' // Excludes users
  | 'seo' // Hurts discoverability
  | 'brand' // Damages perception
  | 'other';

export interface StructuredImprovement {
  id: string;
  issue: string; // What's the problem (required, min 10 chars)
  location?: string; // Where in the work (optional but encouraged)
  suggestion: string; // Concrete fix (required, min 10 chars)
  priority: FeedbackPriority; // How urgent
  // Premium fields for expert reviews
  effort?: EffortEstimate; // How much work to implement
  confidence?: ConfidenceLevel; // How confident in this suggestion
  category?: ImprovementCategory; // Categorize the improvement
  isQuickWin?: boolean; // Flag for quick wins (high impact, low effort)
  resources?: ResourceLink[]; // Supporting links/references
  // NEW: Insight fields that transform opinions into expert guidance
  principle?: string; // What rule/heuristic is being violated (e.g., "Nielsen's visibility of system status")
  principleCategory?: PrincipleCategory; // Category of the principle
  impact?: string; // What happens if not fixed (e.g., "Users abandon at checkout due to confusion")
  impactType?: ImpactType; // Type of impact
  afterState?: string; // What it would look like if fixed (visualization)
}

export interface StructuredStrength {
  id: string;
  what: string; // What's good (required, min 10 chars)
  why?: string; // Why it works well (optional but encouraged)
  // Premium field
  impact?: string; // Business/UX impact of this strength
}

// Expert review premium sections
export interface ExecutiveSummary {
  tldr: string; // 1-3 sentence takeaway (50-200 chars)
  keyStrengths: string[]; // Top 3 bullet points
  keyActions: string[]; // Top 3 priority actions
  overallReadiness?: 'ready' | 'almost-ready' | 'needs-work' | 'major-revision'; // Overall assessment
}

export interface FollowUpOffer {
  available: boolean; // Is reviewer offering follow-up?
  type?: 'code-review' | 'design-feedback' | 'consultation' | 'pair-session' | 'other';
  description?: string; // What they're offering
  responseTime?: string; // Expected response time
}

// Top 3 Takeaways - the most critical actionable items
export interface TopTakeaway {
  issue: string; // Brief description of the issue
  fix: string; // Concrete action to take
}

export interface Phase3DetailedFeedback {
  strengths: string[]; // Legacy: simple strings (1-10 items)
  improvements: string[]; // Legacy: simple strings (1-10 items)
  structured_strengths?: StructuredStrength[]; // New: structured format
  structured_improvements?: StructuredImprovement[]; // New: structured format
  additional_notes?: string; // Optional, up to 5000 chars
  visual_annotations?: VisualAnnotation[]; // Optional visual annotations for design/art (max 20)
  voice_memo?: VoiceMemo; // Optional voice memo
  // Required: Top 3 Takeaways (the TL;DR checklist)
  top_takeaways?: TopTakeaway[]; // 3 most important action items
  // Premium expert review sections
  executive_summary?: ExecutiveSummary; // TL;DR for busy creators
  follow_up_offer?: FollowUpOffer; // Continued support offer
}

// ===== Quality Metrics =====

export interface QualityMetrics {
  completeness_score: number; // 0-100
  estimated_tone: 'professional' | 'casual' | 'critical' | 'encouraging';
  clarity_score: number; // 0-100
  actionability_score: number; // 0-100
}

// ===== Metadata =====

export interface SmartReviewMetadata {
  version: string; // "1.0"
  created_at: string; // ISO date string
  last_updated_at: string; // ISO date string
  time_spent_seconds: number;
  phases_completed: string[]; // ['phase1', 'phase2', 'phase3']
}

// ===== Complete Smart Review Draft =====

export interface SmartReviewDraft {
  phase1_quick_assessment?: Phase1QuickAssessment;
  phase2_rubric?: Phase2RubricRatings;
  phase3_detailed_feedback?: Phase3DetailedFeedback;
  quality_metrics?: QualityMetrics;
  metadata?: SmartReviewMetadata;
}

// ===== Submit Payload =====

export interface SmartReviewSubmit {
  smart_review: SmartReviewDraft;
  attachments?: Array<{
    file_url: string;
    file_name: string;
    file_type: string;
  }>;
}

// ===== Rubric Configuration (from backend) =====

export interface FocusArea {
  id: string;
  label: string;
  description: string;
}

export interface RatingDimension {
  id: string;
  label: string;
  description: string;
  criteria: string[];
}

export interface SectionPrompt {
  id: string;
  label: string;
  prompt: string;
  placeholder: string;
  required: boolean;
  min_items: number;
}

export interface ContentRubric {
  content_type: string;
  focus_areas: FocusArea[];
  rating_dimensions: RatingDimension[];
  section_prompts: Record<string, SectionPrompt>;
}

// ===== UI State Types =====

export type PhaseNumber = 1 | 2 | 3;

export interface PhaseValidation {
  isValid: boolean;
  errors: string[];
}

export interface ReviewEditorState {
  currentPhase: PhaseNumber;
  draft: SmartReviewDraft;
  rubric: ContentRubric | null;
  saveStatus: 'idle' | 'saving' | 'saved' | 'error';
  lastSaved: Date | null;
  qualityMetrics: QualityMetrics | null;
}

// ===== Helper Types =====

export interface PhaseConfig {
  number: PhaseNumber;
  title: string;
  description: string;
  required: boolean;
}

export const PHASE_CONFIGS: Record<PhaseNumber, PhaseConfig> = {
  1: {
    number: 1,
    title: 'Quick Assessment',
    description: 'Overall rating and focus areas',
    required: true,
  },
  2: {
    number: 2,
    title: 'Detailed Ratings',
    description: 'Rate specific dimensions',
    required: true,
  },
  3: {
    number: 3,
    title: 'Detailed Feedback',
    description: 'Strengths and improvements',
    required: false,
  },
};

File: backend/app/schemas/review_slot/smart_review.py
"""Smart Adaptive Review Editor schemas"""

from datetime import datetime
from typing import Optional, List, Any
from pydantic import BaseModel, Field, field_validator

from .feedback import (
    RatingRationale,
    RatingRationaleDraft,
    StructuredStrengthDraft,
    StructuredImprovementDraft,
    TopTakeawayDraft,
    ExecutiveSummaryDraft,
    FollowUpOfferDraft,
)


class Phase1QuickAssessment(BaseModel):
    """Phase 1: Quick assessment with overall rating and focus areas"""
    overall_rating: int = Field(..., ge=1, le=5, description="Overall rating (1-5 stars)")
    primary_focus_areas: List[str] = Field(
        ...,
        min_length=1,
        max_length=6,
        description="Selected focus areas (e.g., ['functionality', 'security'])"
    )
    quick_summary: str = Field(
        ...,
        min_length=50,
        max_length=300,
        description="Brief summary (50-300 chars)"
    )

    @field_validator('quick_summary')
    @classmethod
    def validate_summary(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError('Summary cannot be empty')
        return v


class Phase1QuickAssessmentDraft(BaseModel):
    """Phase 1 Draft: Relaxed validation for saving incomplete drafts"""
    overall_rating: Optional[int] = Field(None, ge=0, le=5, description="Overall rating (0-5, 0 = not set)")
    primary_focus_areas: Optional[List[str]] = Field(
        default=[],
        max_length=6,
        description="Selected focus areas"
    )
    quick_summary: Optional[str] = Field(
        default="",
        max_length=300,
        description="Brief summary (can be incomplete in draft)"
    )


class Phase2RubricRatings(BaseModel):
    """Phase 2: Content-specific rubric ratings"""
    content_type: str = Field(..., description="Content type: code, design, writing")
    ratings: dict[str, int] = Field(
        ...,
        description="Dimension ratings (e.g., {'functionality': 5, 'code_quality': 4})"
    )
    rationales: Optional[dict[str, RatingRationale]] = Field(
        None,
        description="Justification for each rating"
    )

    @field_validator('ratings')
    @classmethod
    def validate_ratings(cls, v: dict[str, int]) -> dict[str, int]:
        """Validate all ratings are 1-5"""
        for dimension, rating in v.items():
            if not (1 <= rating <= 5):
                raise ValueError(f'Rating for {dimension} must be between 1 and 5')
        return v


class Phase2RubricRatingsDraft(BaseModel):
    """Phase 2 Draft: Relaxed validation for saving incomplete drafts"""
    content_type: Optional[str] = Field(None, description="Content type: code, design, writing")
    ratings: Optional[dict[str, int]] = Field(
        default={},
        description="Dimension ratings (can be partial in draft)"
    )
    rationales: Optional[dict[str, RatingRationaleDraft]] = Field(
        default={},
        description="Justification for each rating (can be partial in draft)"
    )


class VisualAnnotation(BaseModel):
    """Visual annotation pin-point on an image"""
    id: str = Field(..., description="Unique annotation ID (client-generated UUID)")
    x: float = Field(..., ge=0, le=100, description="X coordinate as percentage (0-100)")
    y: float = Field(..., ge=0, le=100, description="Y coordinate as percentage (0-100)")
    comment: str = Field(
        ...,
        min_length=1,
        max_length=500,
        description="Comment for this annotation point"
    )


class Phase3DetailedFeedback(BaseModel):
    """Phase 3: Detailed feedback with strengths and improvements"""
    strengths: List[str] = Field(
        ...,
        min_length=1,
        max_length=10,
        description="List of strengths (2-10 items recommended)"
    )
    improvements: List[str] = Field(
        ...,
        min_length=1,
        max_length=10,
        description="List of improvements (2-10 items recommended)"
    )
    additional_notes: Optional[str] = Field(
        None,
        max_length=5000,
        description="Additional notes or context (optional)"
    )
    visual_annotations: Optional[List[VisualAnnotation]] = Field(
        None,
        max_length=20,
        description="Visual annotations for design/art reviews (optional, max 20)"
    )

    @field_validator('strengths', 'improvements')
    @classmethod
    def validate_items(cls, v: List[str]) -> List[str]:
        """Validate list items are not empty"""
        validated = []
        for item in v:
            stripped = item.strip()
            if stripped:
                validated.append(stripped)
        if not validated:
            raise ValueError('List cannot be empty')
        return validated


class Phase3DetailedFeedbackDraft(BaseModel):
    """Phase 3 Draft: Relaxed validation for saving incomplete drafts"""
    strengths: Optional[List[str]] = Field(
        default=[],
        max_length=10,
        description="List of strengths (can be empty in draft)"
    )
    improvements: Optional[List[str]] = Field(
        default=[],
        max_length=10,
        description="List of improvements (can be empty in draft)"
    )
    structured_strengths: Optional[List[StructuredStrengthDraft]] = Field(
        default=None,
        description="Structured strength items with detailed fields"
    )
    structured_improvements: Optional[List[StructuredImprovementDraft]] = Field(
        default=None,
        description="Structured improvement items with detailed fields"
    )
    additional_notes: Optional[str] = Field(
        None,
        max_length=5000,
        description="Additional notes or context (optional)"
    )
    visual_annotations: Optional[List[Any]] = Field(
        None,
        max_length=20,
        description="Visual annotations for design/art reviews"
    )
    voice_memo: Optional[Any] = Field(
        None,
        description="Voice memo data"
    )
    # Required: Top 3 Takeaways (the TL;DR checklist)
    top_takeaways: Optional[List[TopTakeawayDraft]] = Field(
        default=None,
        max_length=3,
        description="Top 3 most important action items"
    )
    # Premium expert review sections
    executive_summary: Optional[ExecutiveSummaryDraft] = Field(
        None,
        description="Executive summary for premium reviews"
    )
    follow_up_offer: Optional[FollowUpOfferDraft] = Field(
        None,
        description="Follow-up offer for premium reviews"
    )


class QualityMetrics(BaseModel):
    """Auto-calculated quality metrics"""
    completeness_score: int = Field(..., ge=0, le=100, description="Completeness (0-100%)")
    estimated_tone: str = Field(..., description="Tone: professional, casual, critical, encouraging")
    clarity_score: int = Field(..., ge=0, le=100, description="Clarity (0-100%)")
    actionability_score: int = Field(..., ge=0, le=100, description="Actionability (0-100%)")


class SmartReviewMetadata(BaseModel):
    """Metadata about the review process"""
    version: str = Field(default="1.0", description="Schema version")
    created_at: datetime
    last_updated_at: datetime
    time_spent_seconds: int = Field(..., ge=0, description="Time spent on review")
    phases_completed: List[str] = Field(
        ...,
        description="Completed phases: ['phase1', 'phase2', 'phase3']"
    )


class SmartReviewDraft(BaseModel):
    """Draft for Smart Adaptive Review Editor - uses relaxed validation for saving incomplete drafts"""
    phase1_quick_assessment: Optional[Phase1QuickAssessmentDraft] = None
    phase2_rubric: Optional[Phase2RubricRatingsDraft] = None
    phase3_detailed_feedback: Optional[Phase3DetailedFeedbackDraft] = None
    quality_metrics: Optional[QualityMetrics] = None
    metadata: Optional[SmartReviewMetadata] = None

    class Config:
        json_schema_extra = {
            "example": {
                "phase1_quick_assessment": {
                    "overall_rating": 4,
                    "primary_focus_areas": ["functionality", "code_quality"],
                    "quick_summary": "Well-structured code with good separation of concerns. Some performance optimizations needed."
                },
                "phase2_rubric": {
                    "content_type": "code",
                    "ratings": {
                        "functionality": 5,
                        "code_quality": 4,
                        "security": 4,
                        "test_coverage": 3
                    }
                },
                "phase3_detailed_feedback": {
                    "strengths": [
                        "Clear function names and good documentation",
                        "Proper error handling throughout"
                    ],
                    "improvements": [
                        "Add unit tests for edge cases",
                        "Consider memoization for expensive calculations"
                    ],
                    "additional_notes": "Overall excellent work. Focus on test coverage for production readiness."
                }
            }
        }


class SmartReviewSubmit(BaseModel):
    """Submit review with Smart Adaptive Review Editor structure"""
    smart_review: SmartReviewDraft
    attachments: Optional[List[dict]] = Field(default=None)

    class Config:
        json_schema_extra = {
            "example": {
                "smart_review": {
                    "phase1_quick_assessment": {
                        "overall_rating": 4,
                        "primary_focus_areas": ["functionality", "code_quality"],
                        "quick_summary": "Well-structured code with some improvements needed."
                    },
                    "phase2_rubric": {
                        "content_type": "code",
                        "ratings": {"functionality": 5, "code_quality": 4}
                    },
                    "phase3_detailed_feedback": {
                        "strengths": ["Clear naming", "Good error handling"],
                        "improvements": ["Add tests", "Optimize performance"]
                    }
                },
                "attachments": []
            }
        }
